libcudacxx/include/cuda/std/detail/libcxx/include/__string

```diff
@@ -0,0 +1,1099 @@
+
+#include <cuda/std/detail/libcxx/include/__assert> // all public C++ headers provide the assertion handler
+#include <cuda/std/detail/libcxx/include/__debug>
+
+#include <cuda/std/__fwd/string.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+// char_traits
+
+template <class _CharT>
+struct _LIBCUDACXX_TEMPLATE_VIS char_traits
+{
+    typedef _CharT    char_type;
+    typedef int       int_type;
+    typedef streamoff off_type;
+#ifndef _LIBCUDACXX_HAS_NO_WCHAR_H
+    typedef streampos pos_type;
+    typedef mbstate_t state_type;
+#endif // !_LIBCUDACXX_HAS_NO_WCHAR_H
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline void _CCCL_CONSTEXPR_CXX14
+        assign(char_type& __c1, const char_type& __c2) noexcept {__c1 = __c2;}
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr bool eq(char_type __c1, char_type __c2) noexcept
+        {return __c1 == __c2;}
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr bool lt(char_type __c1, char_type __c2) noexcept
+        {return __c1 < __c2;}
+
+    _LIBCUDACXX_INLINE_VISIBILITY static _CCCL_CONSTEXPR_CXX14
+    int compare(const char_type* __s1, const char_type* __s2, size_t __n);
+    _LIBCUDACXX_INLINE_VISIBILITY static _CCCL_CONSTEXPR_CXX14
+    size_t length(const char_type* __s);
+    _LIBCUDACXX_INLINE_VISIBILITY static _CCCL_CONSTEXPR_CXX14
+    const char_type* find(const char_type* __s, size_t __n, const char_type& __a);
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static char_type*       move(char_type* __s1, const char_type* __s2, size_t __n);
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static char_type*       copy(char_type* __s1, const char_type* __s2, size_t __n);
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static char_type*       assign(char_type* __s, size_t __n, char_type __a);
+
+#ifndef __cuda_std__
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr int_type  not_eof(int_type __c) noexcept
+        {return eq_int_type(__c, eof()) ? ~eof() : __c;}
+#endif // !__cuda_std__
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr char_type to_char_type(int_type __c) noexcept
+        {return char_type(__c);}
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr int_type  to_int_type(char_type __c) noexcept
+        {return int_type(__c);}
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr bool      eq_int_type(int_type __c1, int_type __c2) noexcept
+        {return __c1 == __c2;}
+#ifndef __cuda_std__
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static inline constexpr int_type  eof() noexcept
+        {return int_type(EOF);}
+#endif
+};
+
+template <class _CharT>
+_CCCL_CONSTEXPR_CXX14 int
+char_traits<_CharT>::compare(const char_type* __s1, const char_type* __s2, size_t __n)
+{
+    for (; __n; --__n, ++__s1, ++__s2)
+    {
+        if (lt(*__s1, *__s2))
+            return -1;
+        if (lt(*__s2, *__s1))
+            return 1;
+    }
+    return 0;
+}
+
+template <class _CharT>
+inline
```

Comment:
Question: I am wondering is the `inline` here needed given that it is already a template?

Also why is none of those functions marked as `_LIBCUDACXX_INLINE_VISIBILITY`?

If they are currently not used I would strongly prefer to just drop them

Reply:
One of the compilers (ICC? I'd need to check) was very unhappy if these were not marked as inline with the `inline` keyword.

They are host-device, they are marked in the class definition. I can move them there if that will help with the readability of this; I think that'd _also_ get rid of the warnings I mentioned above?

I'd also prefer to keep them, because otherwise I'll just have to undo their deletion when I come back here to do string_view (at which point I'll probably mark char_traits as fully supported too).